### ISSUE TITLE
eta-reducing to merge

### ISSUE THREAD

**Issue by imuli** [dhall-lang/dhall-lang#1358]
I often find myself writing a lambda like this:

```dhall
let Example = < A | B >
\(x : Example) -> merge { A = False, B = True } x
```

Especially in a nested union situation, e.g.:

```dhall
merge {
  Nothing = False,
  Some = \(x : Example) -> merge { A = False, B = True } x
}
```

It seems to be a syntax thing that prevents simply writing `merge { A = False, B = True }` in these circumstances. I can't say I have the whole syntax cleanly in my head, but it seems like having `merge {...}` be treated very much like a lambda could work out without other syntax (or semantics) changes.

**Comment by MonoidMusician** [comment]
The reason this isn't allowed is due to the way Dhall infers types: Dhall needs to know what type `x` has before it can determine if the `merge` is well typed. It can't infer what type `x` should have from the record that was passed to it. (Especially since enums are supported.)

It would technically be possible to have an annotated syntax, like for list literals, but you really aren't gaining much at that point (and now you have to specify the return type too):
```dhall
merge { A = False, B = True } : Example -> Bool
```

**Comment by Gabriella439** [comment]
In particular, it's dhall's support for empty alternatives that makes not possible to infer the type of the argument.  To see why, consider the following two dhall expressions:

```dhall
λ(x : < A | B >) → merge { A = Natural/odd, B = Natural/even } x
```

```dhall
λ(x : < A : Natural | B : Natural >) → merge { A = Natural/odd, B = Natural/even } x
```

Both of those expressions are well-typed, and both of those expressions have the same record of handlers, but if you were to η-reduce both of them then dhall wouldn't be able to infer the type of the argument (without a type annotation).

**Comment by imuli** [comment]
There are plenty of cases (especially within other merges) where the type of `x` is fixed by the external context.

```dhall
let Pink = < Fuchsia | Magenta >
let Colour = < Pink : Pink | Red >
in \(x: Colour) -> merge {
  Pink = merge {
    Fuchsia = Natural/even,
    Magenta = Natural/odd,
  },
  Red = Natural/even,
}
```

there's no ambiguity about what the input type to the inner merge. In cases that are ambiguous, I'd imagine you'd need to provide a wrapping lambda as is done currently (or annotate).

**Comment by Gabriella439** [comment]
Yeah, but dhall's language standard is currently not implemented as a bidirectional type checker, which is why it can't (yet) make use of that sort of information
